Automattic/woocommerce-payments
# Add Woo Subscription Stripe Billing settings and migration status notices

Fixes #6508

### Changes proposed in this Pull Request

This PR adds a new "Subscriptions" setting section to the WooPayments setting screen. This new section is only shown if you have the Woo Subscriptions plugin active. 

It contains a simple checkbox to enable/disable Stripe Billing along with an array of notices that are shown depending on the Stripe Billing migration status. 

<p align="center">
<img width="500" alt="Screenshot 2023-09-06 at 1 20 50 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/a9151070-717a-4d9e-9cf0-569a310dc4b2">
</p>

### Testing

<details><summary><strong>No Settings</strong></summary>
<p>

**Setup:**

1. No Woo Subscriptions plugin installed. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Note there isn't any **"Subscriptions"** and Stripe Billing setting shown. 

</p>
</details>

<details><summary><strong>No Stripe Billing subscriptions</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. No pre-existing Stripe Billing subscriptions. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Note there is a new **"Subscriptions"** section in the **Advanced** section with a single setting.
   - The setting should be disabled by default. 
3. Enable/disabling the setting and saving should set the `_wcpay_feature_stripe_billing` option to `1`/`0`.

<img width="725" alt="Screenshot 2023-09-06 at 2 48 13 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/6be5e940-72fc-4faa-9b7d-20d5b063a15e">

</p>
</details>

<details><summary><strong>Enable Stripe Billing</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. A subscription product of any type. 

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
3. Enable the Stripe Billing setting.
4. Purchase a subscription. 
5. Check that the subscription is set up with Stripe Billing.
    - Go to **WooCommerce → Subscriptions**
    - Edit the subscription. 
    - You should notice it has a WCPay subscription ID.
6. This subscription should behave like a Stripe Billing subscription.

<img width="475" alt="Screenshot 2023-08-31 at 3 06 00 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/97f1a979-5087-40f0-aacf-ef7241bc70bd">

</p>
</details>

<details><summary><strong>Disable Stripe Billing with Subscriptions</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions plugin installed. 
2. Stripe Billing enabled.
3. At least 1 Stripe Billing subscription from:
   - A previous test,
   - Created with WCPay Subscriptions built in functionality, or
   - The test generator - see tools on the project thread peNR48-2u-p2

**Testing instructions:**

1. Go to **WooCommerce → Settings → Payments → Woo Payments**
2. Disable the Stripe Billing setting.
4. You should see a warning notice pop up explaining that Subscriptions will be migrated.

<img width="698" alt="Screenshot 2023-09-06 at 2 52 26 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/ab87b29c-8d29-4db1-98a8-f46694c2cf61">

5. Save settings.
6. A new notice should appear indicating that the migration is in-progress. 

<img width="699" alt="Screenshot 2023-09-06 at 2 54 17 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/b46c9224-d702-417e-92a4-f028b59a8ed0">

7. Wait for the migration to complete.
    - if you refresh the settings page while the migration is in progress you'll continue to see that notice. 
    - You can track the migration on the **Tools → Scheduled Actions** screen. Search for `wcpay_schedule_subscription_migrations` and `wcpay_migrate_subscription` actions. 
9. Refresh the settings page and it should so the migration complete notice. 

<img width="705" alt="Screenshot 2023-09-06 at 3 00 10 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/84c16c19-afd6-4cb2-bff8-34b8c6255244">

</p>
</details>

<details><summary><strong>Choose migrate option</strong></summary>
<p>

**Setup:**

1. Woo Subscriptions inactive.
2. WCPay Subscriptions feature enabled. 
4. Delete the `_wcpay_feature_stripe_billing` option we want to simulate a merchant that has upgraded from WCPay Subscriptions to Woo Subscriptions for the first time.

**Testing instructions:**

1. Purchase a subscription (this will be a Stripe Billing subscription).
1. Enable the Woo Subscriptions plugin.
3. Go to **WooCommerce → Settings → Payments → Woo Payments**
4. You should see the following in the Advanced section.

<img width="709" alt="Screenshot 2023-09-06 at 3 07 00 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/a7863ef9-c632-45bb-bad4-94e38a91369c">

5. Click the button to begin migration. 
6. You should see the in-progress notice.

<img width="713" alt="Screenshot 2023-09-06 at 3 08 14 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/1e551ed1-3023-44e3-b683-6aeb42bdbbf4">

7. Wait for the migration to complete. Refresh. 
9. You will see a completed notice. 

<img width="693" alt="Screenshot 2023-09-06 at 3 11 34 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/40285686-b767-4264-aedf-6f9fe9ac8429">

</p>
</details>

There's a lot of depth to these settings and the notices that are shown. They aren't all described above. 


-------------------


**Post merge**
